src/NHibernate/Impl/SessionFactoryImpl.cs

```diff
@@ -105,23 +105,23 @@ public void HandleEntityNotFound(string entityName, string propertyName, object
 			new ConcurrentDictionary<string, CacheBase>();
 
 		[NonSerialized]
-		private readonly IDictionary<string, IClassMetadata> classMetadata;
+		private readonly IReadOnlyDictionary<string, IClassMetadata> classMetadata;
```

Comment:
All theses changes look superfluous to me, because this will never be enough to get the build of a session factory threadsafe with concurrent uses of the factory being built. The only reasonable fix is to not end up using a factory that is concurrently being built.